### PR TITLE
added classic mlst module

### PR DESCRIPTION
**Pull request by lskatz**
## PR checklist

Closes #XXX 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.  _I think I did this right?_
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by lskatz** [comment]
I'm not sure if this is helpful/informative or not 

```(nf-core) [gzu2@monolith3 nf-core-modules]$ nf-core modules lint --dir .

                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 2.1



? Lint all modules or a single named module? Named module
? Tool name: mlst
INFO     Linting modules repo: '.'                                                                   __init__.py:157
INFO     Linting module: 'mlst'                                                                      __init__.py:161
╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Module lint results                                                                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ [!] 3 Test Warnings                                                                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭──────────────────────────────────────────┬───────────────────────────┬───────────────────────────────────────────╮
│ Module name                              │ File path                 │ Test message                              │
├──────────────────────────────────────────┼───────────────────────────┼───────────────────────────────────────────┤
│ mlst                                     │ modules/mlst/main.nf      │ Could not fetch remote copy, skipping     │
│                                          │                           │ comparison.                               │
│ mlst                                     │ modules/mlst/functions.nf │ Could not fetch remote copy, skipping     │
│                                          │                           │ comparison.                               │
│ mlst                                     │ modules/mlst/meta.yml     │ Could not fetch remote copy, skipping     │
│                                          │                           │ comparison.                               │
╰──────────────────────────────────────────┴───────────────────────────┴───────────────────────────────────────────╯
╭───────────────────────╮
│ LINT RESULTS SUMMARY  │
├───────────────────────┤
│ [✔]  22 Tests Passed  │
│ [!]   3 Test Warnings │
│ [✗]   0 Test Failed   │
╰───────────────────────╯
```

File: modules/mlst/functions.nf [hunk 1]
```diff
@@ -0,0 +1,68 @@
+//
+//  Utility functions used in nf-core DSL2 module files
+//
+
+//
```
Comment (drpatelh) [inline modules/mlst/functions.nf:5]: This file will need to be replaced with [this one](https://github.com/nf-core/tools/blob/dev/nf_core/module-template/modules/functions.nf) as a result of the recent changes we have made to be able to record multiple software versions per module.

File: modules/mlst/main.nf [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    path "*.version.txt"          , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    mlst \\
+      --threads $task.cpus \\
+      $fasta \\
+      > ${fasta}.mlst.tsv
```
Comment (drpatelh) [inline modules/mlst/main.nf:35]:
```suggestion
      > ${prefix}.tsv
```

Comment (drpatelh) [reply modules/mlst/main.nf:35]: This will name the output file by sample rather than by the name of the fasta.

File: modules/mlst/main.nf [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    //samtools \\
+    //    sort \\
+    //    $options.args \\
+    //    -@ $task.cpus \\
+    //    -o ${prefix}.bam \\
+    //    -T $prefix \\
+    //    $bam
+
+    //echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' > ${software}.version.txt
```
Comment (drpatelh) [inline modules/mlst/main.nf:53]:
```suggestion
```

File: modules/mlst/main.nf [hunk 4]
```diff
@@ -0,0 +1,55 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
```
Comment (drpatelh) [inline modules/mlst/main.nf:2]:
```suggestion
include { initOptions; saveFiles; getSoftwareName; getProcessName } from './functions'
```

Comment (drpatelh) [reply modules/mlst/main.nf:2]: Updated in line with systematic changes in #665

File: modules/mlst/main.nf [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    } else {
+        container "quay.io/biocontainers/mlst:2.19.0--hdfd78af_1"
+    }
+
+    input:
+    tuple val(meta), path(fasta)
+
+    output:
+    tuple val(meta), path("*.tsv"), emit: tsv
+    path "*.version.txt"          , emit: version
```
Comment (drpatelh) [inline modules/mlst/main.nf:26]:
```suggestion
    tuple val(meta), path("*.tsv"), emit: tsv
    path "versions.yml"           , emit: version
```

Comment (drpatelh) [reply modules/mlst/main.nf:26]: Updated in line with systematic changes in #665

File: modules/mlst/main.nf [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
```
Comment (drpatelh) [inline modules/mlst/main.nf:29]:
```suggestion
```

Comment (drpatelh) [reply modules/mlst/main.nf:29]: Updated in line with systematic changes in #665

File: modules/mlst/main.nf [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    mlst \\
+      --threads $task.cpus \\
+      $fasta \\
+      > ${fasta}.mlst.tsv
+
+    mlst --version > ${software}.version.txt
+
+    echo "MLST" > versions.yml
+    echo "  mlst: \$(mlst --version | grep -o '[0-9.]*')" >> versions.yml
+
```
Comment (drpatelh) [inline modules/mlst/main.nf:41]:
```suggestion
    cat <<-END_VERSIONS > versions.yml
    ${getProcessName(task.process)}:
        mlst: \$( mlst --version | grep -o '[0-9.]*' )
    END_VERSIONS
```

Comment (drpatelh) [reply modules/mlst/main.nf:41]: Updated in line with systematic changes in #665

File: modules/mlst/meta.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
+name: mlst
+description: Run Torsten Seemann's classic MLST on a genome assembly
+keywords:
+  - sort
```
Comment (drpatelh) [inline modules/mlst/meta.yml:4]:
```suggestion
  - mlst
```

Comment (drpatelh) [reply modules/mlst/meta.yml:4]: Any more?

File: modules/mlst/meta.yml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
+  - bam:
```
Comment (drpatelh) [inline modules/mlst/meta.yml:35]:
```suggestion
  - tsv:
```

File: tests/modules/mlst/test.yml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+- name: mlst test_mlst
+  command: nextflow run tests/modules/mlst -entry test_mlst -c tests/config/nextflow.config
+  tags:
+    - mlst
+  files:
+    - path: output/mlst/NCTC13799.fna.mlst.tsv
```
Comment (drpatelh) [inline tests/modules/mlst/test.yml:6]:
```suggestion
    - path: output/mlst/test.tsv
```

**Review by drpatelh — CHANGES_REQUESTED** [review]
Looks great @lskatz ! Mainly minor changes as a result of recent updates we are trying to push to nf-core/modules. Sorry, a little bit of bad timing here.

The CI tests will be stalled because have maxed them out updating all of the other modules https://github.com/nf-core/modules/pull/739. In any case, I think it would be a good idea to get that PR in first and then we can come back to get this in after you have made the appropriate changes :)

**Comment by lskatz** [comment]
Thank you!  I made changes locally so that I could get the muscle memory and pushed those changes to this PR.

File: modules/mlst/main.nf [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    tuple val(meta), path("*.tsv"), emit: tsv
+    path "versions.yml"           , emit: version
+
+    script:
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    mlst \\
+      --threads $task.cpus \\
+      $fasta \\
+      > ${fasta}.tsv
```
Comment (rpetit3) [inline modules/mlst/main.nf]: @lskatz I wonder if `${prefix}` (e.g. sample name) might be better then using the fasta name.

**Comment by rpetit3** [comment]
@drpatelh I think we are good here, as long as @lskatz agrees with my switch to `> ${prefix}.tsv` for the result output.

File: modules/mlst/main.nf (hunk 11 above)
Comment (lskatz) [reply modules/mlst/main.nf]: I have made the suggested change

File: modules/mlst/main.nf [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    } else {
+        container "quay.io/biocontainers/mlst:2.19.0--hdfd78af_1"
+    }
+
+    input:
+    tuple val(meta), path(fasta)
+
+    output:
+    tuple val(meta), path("*.tsv"), emit: tsv
+    path "versions.yml"           , emit: version
```
Comment (drpatelh) [inline modules/mlst/main.nf:26]:
```suggestion
    path "versions.yml"           , emit: versions
```

File: modules/mlst/meta.yml [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+      description: Assembly fasta file
+      pattern: "*.{fasta,fa,fna}"
+
+output:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - version:
```
Comment (drpatelh) [inline modules/mlst/meta.yml:31]:
```suggestion
  - versions:
```

File: modules/mlst/meta.yml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
```
Comment (drpatelh) [inline modules/mlst/meta.yml:34]:
```suggestion
      description: File containing software versions
      pattern: "versions.yml"
```

**Review by drpatelh — APPROVED** [review]

**Comment by drpatelh** [comment]
Thanks guys!

**Pull request merged**